File uploading

Bartzi opened the pull request:
Fixes #36 
It is now possible to add attachments to documents, attachments can be uploaded and will be saved in a media folder, attachments can be deleted and downloaded. This PR also includes tests for that functionality.

**Please Note**
The download feature needs the server to use the mod_sendfile extension as the file download first needs to be checked for sufficient permissions and is then handled by the webserver (this is a nice way i think)

janno42 commented:
Please make sure that the tests delete the uploaded files even if an error occurs during execution.

janno42 commented on _1327/documents/templates/attachments.html:
```diff
@@ -0,0 +1,62 @@
+	</div>
+
+	<script>
+		$('.delete-attachment').on('click', function (event) {
```
deleting needs to have a confirmation

janno42 commented on _1327/documents/tests.py:
```diff
@@ -201,3 +203,187 @@ def test_view_permissions(self):
+class TestAttachments(WebTest):
+	"""
+		Tests that test the work with attachment.
+		Informationdocuments are used as testclass. It is assumed that the behaviour is similar with other documenttypes
```
Please use correct capitalization for classes, "behavior", and change the first sentence to "Tests creating, viewing and deleting attachments".

janno42 commented on _1327/documents/tests.py:
```diff
@@ -201,3 +203,187 @@ def test_view_permissions(self):
+			('file', 'test.txt', bytes(tempfile.SpooledTemporaryFile(max_size=10000, prefix='txt', mode='r')))
+		]
+
+		# test that user that has no change permission on a document can not add an attachment
```
user **who**, can not ... **neither** ...

janno42 commented on _1327/documents/tests.py:
```diff
@@ -201,3 +203,187 @@ def test_view_permissions(self):
+									expect_errors=True)
+		self.assertEqual(response.status_code, 403)
+
+		# test that user that is allowed to view the document may not add attachments to it
```
see above

janno42 commented on _1327/documents/tests.py:
```diff
@@ -201,3 +203,187 @@ def test_view_permissions(self):
 			self.assertIn(subclass.VIEW_PERMISSION_NAME, permission_names, msg="All Document subclasses must declare the permission named in VIEW_PERMISSION_NAME")
+
+
+class TestAttachments(WebTest):
```
check whether all of the empty lines all over the file are necessary

janno42 replied:
please replace all "if the user ... he ..." with "if users ... they ..."

janno42 commented on _1327/documents/tests.py:
```diff
@@ -201,3 +203,187 @@ def test_view_permissions(self):
+									expect_errors=True)
+		self.assertEqual(response.status_code, 403)
+
+		# test that member of group that has according permissions is allowed to upload attachments
```
member of group **who**

janno42 commented on _1327/documents/tests.py:
```diff
@@ -201,3 +203,187 @@ def test_view_permissions(self):
+									user=self.group_user)
+		self.assertEqual(response.status_code, 200)
+
+		# test that superuser is allowed to upload attachments and see the corresponding page
```
**to** see

janno42 commented on _1327/documents/tests.py:
```diff
@@ -201,3 +203,187 @@ def test_view_permissions(self):
+
+		response = self.app.post(reverse('documents:delete_attachment'), params=params, expect_errors=True, xhr=True)
+		self.assertEqual(response.status_code, 403,
+						msg="If a user has no permissions he should not be able to delete an attachment")
```
strange indentation?

janno42 commented on _1327/documents/tests.py:
```diff
@@ -201,3 +203,187 @@ def test_view_permissions(self):
+		self.document = InformationDocument(author=self.user, title="test", text="blabla")
+		self.document.save()
+
+		attachment_file = ContentFile("test content of test attachment")
```
could you check against the content below?

janno42 commented on _1327/information_pages/templates/information_pages_base.html:
```diff
@@ -17,6 +17,7 @@
 			<div class="divider"></div>
 			<a class="btn btn-sm btn-sidebar {% if active_page == 'view' %}btn-warning{% else %}btn-default{% endif %}" href="{% url 'information_pages:view_information' document.url_title %}" role="tab">{% trans "View" %}</a>
 			<a class="btn btn-sm btn-sidebar {% if active_page == 'edit' %}btn-warning{% else %}btn-default{% endif %}" href="{% url 'information_pages:edit' document.url_title %}" role="tab">{% trans "Edit" %}</a>
+			<a class="btn btn-sm btn-sidebar {% if active_page == 'attachments' %}btn-warning{% else %}btn-default{% endif %}" href="{% url 'information_pages:attachments' document.url_title %}" role="tab">{% trans "Attachments" %}</a>
```
here only staff users can see the button

Bartzi replied:
sure this is another problem that already has a ticket ;) see #104

janno42 commented:
deleting attachments does not remove the file from the server. is this intended?

janno42 commented:
the attachment link on the attachments page links to ```/information/attachments/test/documents/00/00/FILENAME/``` which doesn't exist

janno42 commented:
downloading a test attachment returned an empty file (0kb) :(

janno42 commented:
The order of attachments must be configurable

janno42 commented:
The filename of the downloaded file should not have an added ```documents-00-00-``` in the beginning

janno42 commented on _1327/documents/templates/attachments.html:
```diff
@@ -0,0 +1,62 @@
+
+	<h2>{% trans "Already uploaded attachments" %}</h2>
+	<div class="col-sm-offset-2 col-sm-9">
+		<table class="table table-striped">
```
this table should include the filename and the date when the file was uploaded

Bartzi commented:
downloading an attachment using the django development server returns a file with 0KB... this is due to the fact that a real webserver should deliver the requested static file after django checked that the user has the permissions he needs. For this I've chosen to use sendfile, which is a mod to Apache and allows doing this. First step is to check whether the user has sufficient permissions to download the requested document and then it sends a response with the "x-sendfile" header set that tells the webserver to deliver the requested file to the user.
The django development server is not capable of understanding this and thus delivers an empty file.

invliD commented:
I don't like the use of sendfile. And I didn't actually find it. Do you mean mod_xsendfile?

Bartzi commented:
Yes I do mean mod_xsendfile ;) But we can also change it... I just implemented it that way because i thought this is the most django way... serving the 'static' file via a webserver but still checking for permissions using django. But I'm open for another option =)

Bartzi commented:
okay I think I've addressed all comments... deleting an attachment now requires the user to confirm that, the download link is fixed, the files are removed from the server, tests will remove all attachments even if they fail, I've changed the comments and I changed the downloading thing... 
I've added a new library 'django-sendfile' (https://github.com/johnsensible/django-sendfile) that helps us handling the whole thing in a nice and easy way which is able to work with different webservers and still works with the original idea ;) I hope you guys like it, too =)

**If you want to try it make sure to update your pip stuff using the requirements.txt file**

janno42 commented:
The filename on the attachment page should be ```file.name``` instead of ```documents/DD/MM/file.name```

janno42 commented:
If the displayname has no file ending it should get the one from the original file

janno42 commented:
Uploading a file now throws an error for me:
```
'NoneType' object has no attribute 'group'
/vagrant/_1327/information_pages/views.py in attachments
	success, form = handle_attachment(request, document)
/vagrant/_1327/documents/utils.py in handle_attachment
	file_type = re.search(r'\.(\w{3})$', instance.file.name).group(1) 
```

janno42 commented:
:+1:
